app/src/main/kotlin/com/wire/android/feature/OAuthUseCase.kt

```diff
@@ -0,0 +1,153 @@
+        tokenExchangeRequest?.let { request ->
+            authorizationService.performTokenRequest(request) { response, exception ->
+                if (exception != null) {
+                    authState = AuthState()
+                    resultHandler(OAuthResult.Failed(exception.toString()))
+                } else {
+                    if (response != null) {
+                        authState.update(response, exception)
+                        resultHandler(OAuthResult.Success(response.idToken.toString()))
+                    } else {
+                        resultHandler(OAuthResult.Failed.EmptyResponse)
+                    }
+                }
+            }
+        } ?: resultHandler(OAuthResult.Failed.Unknown)
+    }
+
+    private fun getAuthorizationRequest() = AuthorizationRequest.Builder(
+        authServiceConfig, CLIENT_ID, ResponseTypeValues.CODE, Uri.parse(URL_AUTH_REDIRECT)
+    ).setCodeVerifier().setScopes(SCOPE_OPENID, SCOPE_EMAIL, SCOPE_PROFILE).build()
+
+    private fun AuthorizationRequest.Builder.setCodeVerifier(): AuthorizationRequest.Builder {
+        val codeVerifier = getCodeVerifier()
+        setCodeVerifier(
+            codeVerifier, getCodeChallenge(codeVerifier), CODE_VERIFIER_CHALLENGE_METHOD
+        )
+        return this
+    }
+
+    @Suppress("MagicNumber")
+    private fun getCodeVerifier(): String {
+        val secureRandom = SecureRandom()
+        val bytes = ByteArray(64)
+        secureRandom.nextBytes(bytes)
+        return Base64.encodeToString(bytes, ENCODING)
+    }
+
+    private fun getCodeChallenge(codeVerifier: String): String {
+        val hash = MESSAGE_DIGEST.digest(codeVerifier.toByteArray())
+        return Base64.encodeToString(hash, ENCODING)
+    }
+
+    sealed class OAuthResult {
+        data class Success(val idToken: String) : OAuthResult()
+        open class Failed(val reason: String) : OAuthResult() {
+            object Unknown : Failed("Unknown")
+            class InvalidActivityResult(reason: String) : Failed(reason)
+            object EmptyResponse : Failed("Empty Response")
+        }
+    }
+
+    companion object {
+        const val OAUTH_ACTIVITY_RESULT_KEY = "OAuthActivityResult"
+
+        const val SCOPE_PROFILE = "profile"
+        const val SCOPE_EMAIL = "email"
+        const val SCOPE_OPENID = "openid"
+
+        const val CLIENT_ID = "338888153072-4fep6tn6k16tmcbhg4nt4lr65pv3avgi.apps.googleusercontent.com"
+        const val CODE_VERIFIER_CHALLENGE_METHOD = "S256"
+        const val MESSAGE_DIGEST_ALGORITHM = "SHA-256"
+        val MESSAGE_DIGEST = MessageDigest.getInstance(MESSAGE_DIGEST_ALGORITHM)
+        const val ENCODING = Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP
+
+        const val URL_AUTHORIZATION = "https://accounts.google.com/o/oauth2/auth"
+        const val URL_TOKEN_EXCHANGE = "https://oauth2.googleapis.com/token"
+        const val URL_AUTH_REDIRECT = "com.wire.android.internal.debug:/oauth2redirect"
```

Comment:
Should we somehow depend on build config to have different redirect to different builds?